src/librustc/hir/mod.rs

```diff
@@ -961,22 +956,24 @@ pub enum Expr_ {
     ExprRepeat(P<Expr>, P<Expr>),
 }
 
-/// The explicit Self type in a "qualified path". The actual
-/// path, including the trait and the associated item, is stored
-/// separately. `position` represents the index of the associated
-/// item qualified with this Self type.
-///
-///     <HirVec<T> as a::b::Trait>::AssociatedItem
-///      ^~~~~     ~~~~~~~~~~~~~~^
-///      ty        position = 3
-///
-///     <HirVec<T>>::AssociatedItem
-///      ^~~~~    ^
-///      ty       position = 0
+/// Optionally `Self`-qualified value/type path or associated extension.
 #[derive(Clone, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug)]
-pub struct QSelf {
-    pub ty: P<Ty>,
-    pub position: usize,
+pub enum QPath {
+    /// Path to a definition, optionally "fully-qualified" with a `Self` type
+    /// if it points to an associated item, e.g. `<Vec<T> as Trait>::method`.
+    Regular(Option<P<Ty>>, P<Path>),
+
+    /// Associated path extension, e.g. `<T>::default` or `<T>::Output`.
+    /// UFCS source paths can desugar into this, with `Vec::new` turning into
+    /// `<Vec>::new`, and `T::X::Y::method` into `<<<T>::X>::Y>::method`,
+    /// the inner nodes being each one `TyPath(QPath::Extend(..))`.
+    Extend(P<Ty>, P<PathSegment>)
```

Comment:
So, I think the key distinction between "Regular" and "Extend" is that `Extend` requires type-based resolution? I guess that's not quite true, since the same holds for `<X as Y>::Bar` -- but it is true that we have to do a method-style lookup to figure out the trait, right? In any case, I think it'd be helpful for this comment to explain what the difference is.